src/main/java/org/opentripplanner/routing/algorithm/raptoradapter/transit/request/PriorityGroupConfigurator.java

```diff
@@ -0,0 +1,116 @@
+package org.opentripplanner.routing.algorithm.raptoradapter.transit.request;
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import org.opentripplanner.routing.algorithm.raptoradapter.transit.cost.grouppriority.TransitPriorityGroup32n;
+import org.opentripplanner.routing.api.request.request.filter.TransitPriorityGroupSelect;
+import org.opentripplanner.transit.model.framework.FeedScopedId;
+import org.opentripplanner.transit.model.network.RoutingTripPattern;
+
+/**
+ * This class dynamically build an index of transit-group-ids from the
+ * provided {@link TransitPriorityGroupSelect}s while serving the caller with
+ * group-ids for each requested pattern. It is made for optimal
+ * performance, since it is used in request scope.
+ * <p>
+ * THIS CLASS IS NOT THREAD-SAFE.
+ */
+public class PriorityGroupConfigurator {
+
+  private static final int BASE_GROUP_ID = TransitPriorityGroup32n.groupId(0);
+  private int groupIndexCounter = 0;
+  private final boolean enabled;
+  private final PriorityGroupMatcher[] baseMatchers;
+  private final PriorityGroupMatcher[] agencyMatchers;
+  private final PriorityGroupMatcher[] globalMatchers;
+  private final Map<PriorityGroupMatcher, Map<FeedScopedId, Integer>> agencyMatchersIds = new HashMap<>();
+  private final Map<PriorityGroupMatcher, Integer> globalMatchersIds = new HashMap<>();
+
+  private PriorityGroupConfigurator() {
+    this.enabled = false;
+    this.baseMatchers = null;
+    this.agencyMatchers = null;
+    this.globalMatchers = null;
+  }
+
+  private PriorityGroupConfigurator(
+    Collection<TransitPriorityGroupSelect> base,
+    Collection<TransitPriorityGroupSelect> byAgency,
+    Collection<TransitPriorityGroupSelect> global
+  ) {
+    this.baseMatchers =
+      base
+        .stream()
+        .map(PriorityGroupMatcher::of)
+        .filter(Predicate.not(PriorityGroupMatcher::isEmpty))
+        .toArray(PriorityGroupMatcher[]::new);
```

Comment:
I think you want to introduce a reusable method to remove the duplication.

Reply:
And use a static import for `not()` to make it more readable.

Reply:
Fixed in 1b8520f